084-templating-host-and-advertisedHost-fields.md

```diff
@@ -0,0 +1,165 @@
+        key: tls.key
+        secretName: my-certificate
+      brokers:
+        - advertisedHost: broker-10.strimzi.gateway.api.test
+          advertisedPort: 9092
+          broker: 10
+        - advertisedHost: broker-11.strimzi.gateway.api.test
+          advertisedPort: 9092
+          broker: 11
+        - advertisedHost: broker-12.strimzi.gateway.api.test
+          advertisedPort: 9092
+          broker: 12
+```
+
+## Motivation
+
+Being required to specify these configurations for each broker has some negative side-effects:
+* The `Kafka` custom resource gets unnecessarily complicated, especially for bigger clusters.
+  This also increases the likelihood of typos and other configuration issues.
+* With the introduction of KRaft and node pools, the broker IDs are not always just 0, 1, 2 etc.
+  The brokers might now have different numbers and might include gaps.
+  The controller nodes share the same node ID space, but their hostnames do not need to be configured.
+  All of these easily lead to confusion of how should this configuration look like.
+* When adding new brokers, user needs to make sure the `host` or `advertisedHost` fields will be added for the new brokers before scaling the cluster up.
+
+But in most cases, the (advertised) hostnames often follow the same pattern.
+In the example above, you can see that the Ingress based listener uses the hostname pattern `my-cluster-kafka-<NODE-ID>.myingress.com`.
+And the customized listener uses the advertised hostname pattern `broker-<NODE-ID>.strimzi.gateway.api.test`.
+
+While there might be users using more random hostnames, for many Strimzi users, having an easy way to template the hostnames as suggested in this proposal would provide a huge simplification:
+* Users will be able to scale-up without changing the configuration first
+* The `Kafka` CR will be much smaller, simpler and easier to read
+* The likelihood of typos and misconfigurations will be smaller
+
+## Proposal
+
+This proposal suggests adding two new fields to the `configuration` section of the listener:
+* `advertisedHostTemplate`
+* `hostTemplate`
+
+These fields will be strings that will be used to configure templates that will be used to generate the hostnames.
+The templates will support several different _variables_:
+* `{nodeId}` will be replaced with the ID of the Kafka node to which the template is applied
+* `{nodePodName}` will be replaced with the name of the Kafka node (name of the Kubernetes Pod used for this node) to which the template is applied
+
+Strimzi will take these template fields and replace the variables with the corresponding value for each Kafka node.
+And the resulting value would be used for the (advertised) hostnames.
+
+The templates will be applied only to the per-broker values.
+The hostname for the bootstrap Ingress resource would still need to be specified directly and will never be generated from the template.
+(Advertised hostnames can be used only as per-broker configuration)
```

Comment:
I don't think the suggestion makes sense here ... in particular the `This is because` part.
* For the `host` field, the template is not used for the bootstrap `host` because the bootstrap `host` would use a different format/pattern. You cannot apply the template such as `broker-{nodeId}.myingress.com` to the bootstrap as there is no node ID. And you also have only one bootstrap `host`, so having a separate template makes no sense either.
* For the `advertisedHostname`, that is not supported at the bootstrap level because it is a configuration inside the Kafka configuration and can exist only per broker. (Whereas the `host` configuration is on the Kubernetes level)

Maybe we could use something like this?

```suggestion
The templates will only be applied to per-broker values. 
The hostname for the bootstrap Ingress resource must be specified manually and will not be generated from the template. 
The advertised hostnames are used only for per-broker configurations and cannot be configured for bootstrap.
```

Reply:
works for me :+1: